sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -265,6 +265,8 @@ case class AdaptiveSparkPlanExec(
                 } else {
                   events.offer(StageFailure(stage, res.failed.get))
                 }
+                // explicitly clean up the resources in this stage
+                stage.cleanupResources()
```

Comment:
shall we do cleanup for final stage ?  and also for some plans which only have one stage

Reply:
good point, I think we should cleanup for final stage

for single plan without shuffle, I am not very sure right now

Reply:
The execution of the final stage is out of AQE's control. e.g. people can even do `df.rdd` and we don't know when the job of the last job will be submitted.